File: C#/LocalSearchOptimization/Examples/Structures/Tree/EmptyLeafMove.cs
using System.Collections.Generic;
using LocalSearchOptimization.Components;

namespace LocalSearchOptimization.Examples.Structures.Tree
{
    public class EmptyLeafMove : Operator
    {
        public EmptyLeafMove(int nodesNumber, double weight = 1) : base(nodesNumber, weight)
        {
            for (int i = 0; i < 2 * nodesNumber; i++)
                for (int j = 0; j < 2 * nodesNumber - 1; j++)
                    Configurations.Add(new TwoOperands(i, j, this));
        }

        public override ISolution Apply(ISolution solution, Configuration configuration)
        {
            ITreeBranching tree = (ITreeBranching)solution;
            TwoOperands operands = (TwoOperands)configuration;

            List<bool> branching = tree.Branching;

            int start = operands.First;
            int traverse = branching[start] == false ? 1 : -1;

            while (branching[start] == branching[start + traverse])
                start += traverse;

            int left = traverse > 0 ? start : start - 1;

            if (left != operands.Second)
            {
                branching = new List<bool>(tree.Branching);
                branching.RemoveRange(left, 2);
                branching.Insert(operands.Second, true);
                branching.Insert(operands.Second, false);
            }

            return tree.FetchBranching(branching, "empty leaf");
        }
    }
}

File: java/src/localsearchoptimization/examples/structures/tree/EmptyLeafMove.java
package localsearchoptimization.examples.structures.tree;

import localsearchoptimization.components.Configuration;
import localsearchoptimization.components.Operator;
import localsearchoptimization.components.Solution;
import localsearchoptimization.examples.structures.TwoOperands;

public class EmptyLeafMove extends Operator {

    public EmptyLeafMove(int nodesNumber) {
        this(nodesNumber, 1);
    }

    public EmptyLeafMove(int nodesNumber, double weight) {
        super(nodesNumber, weight);
        for (int i = 0; i < 2 * nodesNumber; i++)
            for (int j = 0; j < 2 * nodesNumber - 1; j++)
                configurations.add(new TwoOperands(i, j, this));
    }

    @Override
    public Solution apply(Solution solution, Configuration configuration) {
        TreeBranching tree = (TreeBranching) solution;
        TwoOperands operands = (TwoOperands) configuration;

        boolean[] branching = tree.getBranching();

        int start = operands.first;
        int traverse = branching[start] == false ? 1 : -1;

        while (branching[start] == branching[start + traverse])
            start += traverse;
        int left = traverse > 0 ? start : start - 1;

        if (left != operands.second) {
            branching = new boolean[tree.getBranching().length];
            branching[operands.second] = false;
            branching[operands.second + 1] = true;
            if (left < operands.second) {
                for (int i = 0; i < left; i++)
                    branching[i] = tree.getBranching()[i];
                for (int i = left + 2; i < operands.second + 2; i++)
                    branching[i - 2] = tree.getBranching()[i];
                for (int i = operands.second + 2; i < branching.length; i++)
                    branching[i] = tree.getBranching()[i];
            } else {
                for (int i = 0; i < operands.second; i++)
                    branching[i] = tree.getBranching()[i];
                for (int i = operands.second; i < left; i++)
                    branching[i + 2] = tree.getBranching()[i];
                for (int i = left + 2; i < branching.length; i++)
                    branching[i] = tree.getBranching()[i];
            }
        }

        return tree.fetchBranching(branching, "empty leaf");
    }
}